### ISSUE TITLE
DLL/SO only found on android with IL2CPP NON-Development build

### ISSUE THREAD

**Issue by SoylentGraham** [NewChromantics/MinimalUnityPlugin#1]
Mono development - can't find assembly
Mono release - can't find assemble
IL2CPP development - works
IL2CPP release - `dlopen` fails ("cant find library")

```
DllNotFoundException: Unable to load DLL 'Minimal'. Tried the load the following dynamic libraries: Unable to load dynamic library 'Minimal' because of 'Failed to open the requested dynamic library (0x06000000) dlerror() = dlopen failed: library "Minimal" not found
```

`Minimal.so`'s appear in APK in every case.

**Comment by SoylentGraham** [comment]
In development mode, `libunity.so` adds these symbols, of which I'm probably not referencing (and I [am forced] to link to libc anyway)
```
U getc@LIBC
U getpeername@LIBC
U getrusage@LIBC
U siglongjmp@LIBC
U ungetc@LIBC
U unw_get_proc_info
U unw_get_reg
U unw_getcontext
U unw_init_local
U unw_set_reg
```

**Comment by SoylentGraham** [comment]
Tried with shared c++ as well as static in the android makefile.
Tried downgrading to c++11
NDK doesn't allow linking to gnu_stl any more.
Removed things like exceptions

**Comment by SoylentGraham** [comment]
`Strip Engine Code` Off (IL2CPP only option?) doesn't fix

**Comment by SoylentGraham** [comment]
`C++ Compiler Config: Release` (default) not working
`C++ Compiler Config: Debug` Not working

**Comment by SoylentGraham** [comment]
Adding `-soname,` to android build.... (recently had trouble with this when linking other so's)...
soname is already present;
`objdump -p Minimal.Android/libs/armeabi-v7a/Minimal.so|grep SONAME`
`SONAME       Minimal.so`

**Comment by SoylentGraham** [comment]
A manual dlopen (see cf4d371) succeeds in a dev build
`08-10 18:22:42.719  7349  7368 I MinimalCpp: dlopen(Minimal.so) success!`
fails in non-dev build

**Comment by SoylentGraham** [comment]
adding `LOCAL_LDFLAGS	+= -rdynamic` to `android.mk` didn't work

**Comment by SoylentGraham** [comment]
`PopH264 1.3.41` DOES load in non-dev
https://github.com/NewChromantics/PopH264/releases/tag/v1.3.41

**Comment by SoylentGraham** [comment]
renaming `Minimal.so` to `libMinimal.so`.... works in non-dev.

with
````
[DllImport("Minimal", CallingConvention = CallingConvention.Cdecl)]
private static extern int	Minimal_GetVersionThousand();
````

**Comment by studentutu** [comment]
you should probably add link.xml inside the unity folder and add into there assemblies you will be using.

**Comment by studentutu** [comment]
also, it is not entirely correct to just leave libs/code as is - Unity has a particular folder called Plugins

Please see https://docs.unity3d.com/Manual/NativePlugins.html 

E.g. : you should add Plugins folder in any runtime folder/subfolder, and it will notify Unity to look for .so/ .dll / .aar / .jar / .h / .m / .mm / .cpp and others

link.xml -> will notify Unity release stripping engine to not strip classes/methods which are specified in there ( you will probably need to set the whole Namespace for your own code, and also code that you will be using from the Unity side)

Also, for the build libraries such as .so - you should build for both platforms - use subfolders x86  or x86_64 respectively

**Comment by SoylentGraham** [comment]
> you should probably add link.xml inside the unity folder and add into there assemblies you will be using.

Why?
Isn't that to prevent symbol stripping? Which is not the problem (and unity won't be stripping from my dynamic lib)

**Comment by SoylentGraham** [comment]
> also, it is not entirely correct to just leave libs/code as is - Unity has a particular folder called Plugins
> Please see https://docs.unity3d.com/Manual/NativePlugins.html
> 
> E.g. : you should add Plugins folder in any runtime folder/subfolder, and it will notify Unity to look for .so/ .dll / .aar / .jar / .h / .m / .mm / .cpp and others

Unity removed this requirement about... 5 years ago?
If you read the link you pasted, you'll see no reference to the Plugins/ folder any more. 
For several years now I've been making plugins without the plugin folder, the only real plugin folder required now is for raw .java files (although I think maybe even that requirement is gone)

**Comment by SoylentGraham** [comment]
> Also, for the build libraries such as .so - you should build for both platforms - use subfolders x86 or x86_64 respectively

You will notice, I already build 4 android ABI's, (including x86 and x64) each in their own folder. (which was irrelevent to the problem)

https://github.com/NewChromantics/MinimalUnityPlugin/tree/main/UnityTestProject/Assets/MinimalLib/Android

**Comment by SoylentGraham** [comment]
TLDR; Unity non-development builds cannot find android shared-objects/dlls that aren't prefixed with `lib`
`libMinimal.so` works
`Minimal.so` does not work (except in IL2CPP development builds)

**Comment by SoylentGraham** [comment]
I am assuming (without looking at old unity5 source code) that the profiler loads all libs it finds in the apk, (maybe to enumerate symbols)
As the `SONAME` is `Minimal.so`, I can imagine if unity then tries to `dlopen("Minimal.so")` the process has already found a match with that `SONAME` and resolves.

Although, I dont think unity would actually look for `Minimal.so`... my `dlopen("Minimal.so")` in my cpp source plugin (in the unity project) could prove that, that is why it succeeds there

**Comment by SoylentGraham** [comment]
Unity issue;  IN-13675 
https://unity3d.atlassian.net/servicedesk/customer/portal/2/IN-13675
